File: ImperatorToCK3Tests/MapperTests/TagTitleMapper/TagTitleMappingTests.cpp
#include "../ImperatorToCK3/Source/Mappers/TagTitleMapper/TagTitleMapping.h"
#include "gtest/gtest.h"
#include <sstream>


TEST(Mappers_TagTitleMappingTests, simpleTagMatch)
{
	std::stringstream mappingStream;
	mappingStream << "{ ck3 = e_roman_empire imp = ROM }";

	const mappers::TagTitleMapping theMapping(mappingStream);
	const auto& match = theMapping.tagRankMatch("ROM", "");

	ASSERT_EQ("e_roman_empire", *match);
}

TEST(Mappers_TagTitleMappingTests, simpleTagMatchFailsOnWrongTag)
{
	std::stringstream mappingStream;
	mappingStream << "{ ck3 = e_roman_empire imp = REM }";

	const mappers::TagTitleMapping theMapping(mappingStream);
	const auto& match = theMapping.tagRankMatch("ROM", "");

	ASSERT_EQ(std::nullopt, match);
}

TEST(Mappers_TagTitleMappingTests, simpleTagMatchFailsOnNoTag)
{
	std::stringstream mappingStream;
	mappingStream << "{ ck3 = e_roman_empire }";

	const mappers::TagTitleMapping theMapping(mappingStream);
	const auto& match = theMapping.tagRankMatch("ROM", "");

	ASSERT_EQ(std::nullopt, match);
}

TEST(Mappers_TagTitleMappingTests, tagRankMatch)
{
	std::stringstream mappingStream;
	mappingStream << "{ ck3 = e_roman_empire imp = ROM rank = e }";

	const mappers::TagTitleMapping theMapping(mappingStream);
	const auto& match = theMapping.tagRankMatch("ROM", "e");

	ASSERT_EQ("e_roman_empire", *match);
}

TEST(Mappers_TagTitleMappingTests, tagRankMatchFailsOnWrongRank)
{
	std::stringstream mappingStream;
	mappingStream << "{ ck3 = e_roman_empire imp = ROM rank = k }";

	const mappers::TagTitleMapping theMapping(mappingStream);
	const auto& match = theMapping.tagRankMatch("ROM", "e");

	ASSERT_EQ(std::nullopt, match);
}

TEST(Mappers_TagTitleMappingTests, tagRankMatchSucceedsOnNoRank)
{
	std::stringstream mappingStream;
	mappingStream << "{ ck3 = e_roman_empire imp = ROM }";

	const mappers::TagTitleMapping theMapping(mappingStream);
	const auto& match = theMapping.tagRankMatch("ROM", "e");

	ASSERT_EQ("e_roman_empire", *match);
}

File: ImperatorToCK3.UnitTests/Mappers/TagTitle/TagTitleMappingTests.cs
using ImperatorToCK3.Mappers.TagTitle;
using commonItems;
using Xunit;

namespace ImperatorToCK3.UnitTests.Mappers.TagTitle {
	public class TagTitleMappingTests {
		[Fact]
		public void SimpleTagMatch() {
			var reader = new BufferedReader("{ ck3 = e_roman_empire imp = ROM }");
			var mapping = TagTitleMapping.Parse(reader);
			var match = mapping.TagRankMatch("ROM", "");

			Assert.Equal("e_roman_empire", match);
		}

		[Fact]
		public void SimpleTagMatchFailsOnWrongTag() {
			var reader = new BufferedReader("{ ck3 = e_roman_empire imp = REM }");
			var mapping = TagTitleMapping.Parse(reader);
			var match = mapping.TagRankMatch("ROM", "");

			Assert.Null(match);
		}

		[Fact]
		public void SimpleTagMatchFailsOnNoTag() {
			var reader = new BufferedReader("{ ck3 = e_roman_empire }");
			var mapping = TagTitleMapping.Parse(reader);
			var match = mapping.TagRankMatch("ROM", "");

			Assert.Null(match);
		}

		[Fact]
		public void TagRankMatch() {
			var reader = new BufferedReader("{ ck3 = e_roman_empire imp = ROM rank = e }");
			var mapping = TagTitleMapping.Parse(reader);
			var match = mapping.TagRankMatch("ROM", "e");

			Assert.Equal("e_roman_empire", match);
		}

		[Fact]
		public void TagRankMatchFailsOnWrongRank() {
			var reader = new BufferedReader("{ ck3 = e_roman_empire imp = ROM rank = k }");
			var mapping = TagTitleMapping.Parse(reader);
			var match = mapping.TagRankMatch("ROM", "e");

			Assert.Null(match);
		}

		[Fact]
		public void TagRankMatchSucceedsOnNoRank() {
			var reader = new BufferedReader("{ ck3 = e_roman_empire imp = ROM }");
			var mapping = TagTitleMapping.Parse(reader);
			var match = mapping.TagRankMatch("ROM", "e");

			Assert.Equal("e_roman_empire", match);
		}
	}
}
